contrib/internal/graphql/graphql.go

```diff
@@ -0,0 +1,173 @@
+type Error struct {
+	OriginalErr error
+	Message     string
+	Locations   []ErrorLocation
+	Path        []any
+	Extensions  map[string]any
+}
+
+// ErrorLocation represents each individual member of the locations field of a Graphql error.
+type ErrorLocation struct {
+	Line   int
+	Column int
+}
+
+// AddErrorsAsSpanEvents attaches the given graphql errors to the span as span events, according to the standard
+// Datadog specification.
+// errExtensions allows to include the given extensions field from the error as attributes in the span events.
+func AddErrorsAsSpanEvents(span tracer.Span, errs []Error, errExtensions []string) {
+	ts := time.Now()
+	for _, err := range errs {
+		ddtrace.AddSpanEvent(
+			span,
+			ext.GraphqlQueryErrorEvent,
+			ddtrace.WithSpanEventTimestamp(ts),
+			ddtrace.WithSpanEventAttributes(errToSpanEventAttributes(err, errExtensions)),
+		)
+	}
+}
+
+func errToSpanEventAttributes(gErr Error, errExtensions []string) map[string]any {
+	res := map[string]any{
+		"message":    gErr.Message,
+		"type":       reflect.TypeOf(gErr.OriginalErr).String(),
+		"stacktrace": takeStacktrace(0, 0),
+	}
+	if locs := parseErrLocations(gErr.Locations); len(locs) > 0 {
+		res["location"] = locs
+	}
+	if errPath := parseErrPath(gErr.Path); len(errPath) > 0 {
+		res["path"] = errPath
+	}
+	setErrExtensions(res, gErr.Extensions, errExtensions)
+	return res
+}
+
+func parseErrLocations(locs []ErrorLocation) []string {
+	res := make([]string, 0, len(locs))
+	for _, loc := range locs {
+		res = append(res, fmt.Sprintf("%d:%d", loc.Line, loc.Column))
+	}
+	return res
+}
+
+func parseErrPath(p []any) []string {
+	res := make([]string, 0, len(p))
+	for _, v := range p {
+		res = append(res, fmt.Sprintf("%v", v))
+	}
+	return res
+}
+
+func setErrExtensions(result map[string]any, extensions map[string]any, whitelist []string) {
+	for _, errExt := range whitelist {
+		val, ok := extensions[errExt]
+		if !ok {
+			continue
+		}
+		key := fmt.Sprintf("extensions.%s", errExt)
+		mapVal, err := errExtensionMapValue(val)
+		if err != nil {
+			log.Debug("failed to set error extension as span event attribute %s: %v", errExt, err)
+			continue
+		}
+		result[key] = mapVal
+	}
+}
+
+func errExtensionMapValue(val any) (any, error) {
+	switch v := val.(type) {
+	case string, bool, int, uint, int64, uint64, uint8, uint16, uint32, uintptr, int8, int16, int32, float64, float32,
+		[]string, []bool, []int, []uint, []int64, []uint64, []uint8, []uint16, []uint32, []uintptr, []int8, []int16, []int32, []float64, []float32:
+		return v, nil
+	default:
+		b, err := json.Marshal(val)
+		if err != nil {
+			return nil, err
+		}
+		return string(b), nil
+	}
+}
+
+// defaultStackLength specifies the default maximum size of a stack trace.
+const defaultStackLength = 32
+
+// takeStacktrace takes a stack trace of maximum n entries, skipping the first skip entries.
+// This function is the same as ddtrace/tracer/span.go
+func takeStacktrace(n, skip uint) string {
```

Comment:
We already have similar helper functions in the code base. Could we use one of the existing ones?